### PR TITLE
fix: accumulate OAuth scopes on 401/403 instead of overwriting

### LINKED ISSUE #1582: Scope overwrite in 403 upscoping prevents progressive authorization for servers with per-operation scopes
**Describe the bug**

The `StreamableHTTPClientTransport` 403 `insufficient_scope` handler overwrites `this._scope` with the scope from the `WWW-Authenticate` header instead of accumulating scopes across responses:

https://github.com/modelcontextprotocol/typescript-sdk/blob/main/packages/client/src/client/streamableHttp.ts#L540

```typescript
if (scope) {
  this._scope = scope; // overwrites - previous scopes are lost
}
```

This causes an infinite re-authorization loop when an MCP server requires different scopes for different operations (progressive/step-up authorization).

The same overwrite exists on https://github.com/modelcontextprotocol/typescript-sdk/blob/main/packages/client/src/client/streamableHttp.ts#L507 in the 401 handler. 

The Python SDK has the same behavior in https://github.com/modelcontextprotocol/python-sdk/blob/main/src/mcp/client/auth/oauth2.py.

To Reproduce

1. Configure an MCP server with per-operation scopes:
  - initialize requires `init` scope
  - tools/list requires `mcp:tools:read` scope
  - tools/call (specific tool) requires `mcp:tools:write`
  
2. When a request lacks the required scope, the server returns 403 with only the scopes that specific operation needs per  https://datatracker.ietf.org/doc/html/rfc6750#section-3.1 (e.g. WWW-Authenticate: Bearer error="insufficient_scope", scope="mcp:tools:read")

3. Connect with the TypeScript SDK using StreamableHTTPClientTransport with an OAuth provider
 
4. After initialize succeeds (granted init), call tools/list
 
5. The 403 handler sets _scope = "mcp:tools:read", re-authorizes, and the auth server grants mcp:tools:read - but init was not requested, so it is not  included in the new token
 
6. The next operation requiring init fails with 403, overwriting _scope = "init", losing mcp:tools:read
 
7. Infinite loop between steps 5 and 6
  
Expected behavior

The transport should accumulate (union) scopes across 401/403 responses. After steps 1–5 above, _scope should be "init mcp:tools:read", not just "mcp:tools:read".

It is incorrect to expect the MCP server to solve this by returning all accumulated scopes in the WWW-Authenticate header because:

- Per RFC 6750 §3.1, the scope attribute describes "the scope necessary to access the protected resource" - the specific resource being accessed, not every resource the client might access in the future.
- The server has no knowledge of client-side token state. It validates the presented token and reports what's missing for this specific operation.
- Different operations have different scope requirements. Including scopes beyond what the operation needs (e.g. returning init in a tools/list 403) would misrepresent the requirements of that operation.

The responsibility for accumulating scopes across operations belongs to the client, not the server.

Related: #1039, #1115, #1151, #941, #1317

### PR DESCRIPTION
**Pull request by rechedev9**
## Summary

- Replaces raw `this._scope = scope` assignments with a `mergeScopes()` utility that unions existing and incoming scope strings (space-separated, `Set`-based deduplication, insertion-order stable)
- Fixes both `StreamableHTTPClientTransport` (lines 520, 553) and `SSEClientTransport` (lines 167, 281)
- Prevents infinite re-authorization loops when an MCP server uses per-operation progressive authorization per [RFC 6750 §3.1](https://datatracker.ietf.org/doc/html/rfc6750#section-3.1)

### Root cause

The 401/403 handlers overwrote `this._scope` with only the scope from the current `WWW-Authenticate` header. With per-operation scopes (e.g., `init` for initialize, `mcp:tools:read` for tools/list), re-authorizing for the new scope dropped coverage of all prior scopes, causing an infinite loop between operations.

### Approach

A module-level unexported `mergeScopes(existing, incoming)` function:
1. Splits both scope strings on whitespace
2. Unions via `Set<string>` (deduplicates, preserves insertion order)
3. Returns `undefined` when the result is empty (matches `_scope?: string` semantics)

Duplicated in both transport files to keep each self-contained — no new shared modules or public API surface.

### Testing

Added 8 regression tests:
- `streamableHttp.test.ts`: mergeScopes edge cases, multi-scope accumulation across 401→403 sequences, circuit-breaker compatibility
- `sse.test.ts`: scope accumulation in EventSource reconnect and send paths

All 265 client tests pass. Typecheck and lint pass.

Fixes #1582

### CODE REVIEW

**Comment by felixweinberger** [comment]
@claude review

File: packages/client/src/client/streamableHttp.ts [hunk 1]
```diff
@@ -16,6 +16,21 @@ import { EventSourceParserStream } from 'eventsource-parser/stream';
 import type { AuthResult, OAuthClientProvider } from './auth.js';
 import { auth, extractWWWAuthenticateParams, UnauthorizedError } from './auth.js';
 
+/**
+ * Merges two space-separated OAuth scope strings into a deduplicated union.
+ * Returns undefined when the resulting set is empty.
+ * Preserves insertion order of first occurrence for determinism.
+ */
+function mergeScopes(existing: string | undefined, incoming: string | undefined): string | undefined {
```
Comment (felixweinberger) [inline packages/client/src/client/streamableHttp.ts]: seems like this should be in `core` rather than defined in 2 places.

**Review by felixweinberger — CHANGES_REQUESTED** [review]

**Comment by rechedev9** [comment]
## Changes in this update

Rebased on `upstream/main` and reworked the implementation + tests:

### Source changes
- **5 merge sites** (vs 4 in the original): also fixed the `_startOrAuthSse` 401 handler at `streamableHttp.ts:240` which was overwriting scopes identically to the other sites
- `mergeScopes()` remains duplicated in each transport file (no shared module), matching the original design intent

### Test improvements
- Refactored `streamableHttp.test.ts` tests to use **403 paths** instead of 401 paths for scope accumulation assertions — the 401 path goes through `handleOAuthUnauthorized` → `auth()` as an internal ESM call, which `vi.spyOn` on the module namespace cannot intercept
- Used **minimal `AuthProvider`** (not `OAuthClientProvider`) in the SSE test to avoid `adaptOAuthProvider` wrapping, allowing direct `UnauthorizedError`-based scope inspection
- Extracted shared test helpers within the `mergeScopes` describe block: `testMessage`, `getScope()`/`setScope()`, `sortedTokens()`, and `beforeEach`/`afterEach` for the auth spy — reducing boilerplate

### Verification
- 319/319 client tests pass
- Typecheck clean

**Comment by felixweinberger** [comment]
@claude review

**Comment by rechedev9** [comment]
Pushed follow-up commit `731ffae`.

Addressed the latest review comments and the adjacent 401 metadata issue:
- preserve `resourceMetadataUrl` when a later 401 omits `resource_metadata`
- add `@modelcontextprotocol/core` to the changeset because `mergeScopes` is now exported there
- make the auth spies restore in `streamableHttp` tests deterministic with `try/finally`
- add regression coverage for repeated 401s in both `streamableHttp` and `sse`

Verified locally and again via the pre-push hook:
- focused client/core tests
- `pnpm -r typecheck`
- `pnpm -r build`
- `pnpm -r lint`

**Comment by felixweinberger** [comment]
@claude review

**Review by felixweinberger — CHANGES_REQUESTED** [review]
Back to your queue to address claude's finding above, looks legit to me.

File: packages/client/src/client/streamableHttp.ts [hunk 2]
```diff
@@ -514,8 +517,10 @@ export class StreamableHTTPClientTransport implements Transport {
...
                         const { resourceMetadataUrl, scope } = extractWWWAuthenticateParams(response);
-                        this._resourceMetadataUrl = resourceMetadataUrl;
-                        this._scope = scope;
+                        if (resourceMetadataUrl) {
+                            this._resourceMetadataUrl = resourceMetadataUrl;
+                        }
+                        this._scope = mergeScopes(this._scope, scope);
                     }
 
                     if (this._authProvider.onUnauthorized && !isAuthRetry) {
```
[the remark this answers is not in the document]
Comment (rechedev9) [reply packages/client/src/client/streamableHttp.ts:527]: Fixed in `7e2cf80`/`3b625e0`.

I threaded `accumulatedScope` through `UnauthorizedContext` and now the 401 interactive OAuth path prefers the merged scope instead of re-reading only the current challenge scope in `handleOAuthUnauthorized()`. I also passed the accumulated scope through all 401 `onUnauthorized(...)` transport call sites in both streamable HTTP and SSE.

Validation:
- `pnpm --filter @modelcontextprotocol/client typecheck`
- `pnpm --filter @modelcontextprotocol/client exec vitest run test/client/auth.test.ts test/client/streamableHttp.test.ts test/client/sse.test.ts`
- pre-push hook also passed full repo build/typecheck/lint during push

**Comment by localden** [comment]
This is also covered in https://github.com/modelcontextprotocol/typescript-sdk/pull/1624 (@SamMorrowDrums) which will cover the SEP-based implementation

**Comment by SamMorrowDrums** [comment]
@localden I need to update that PR really. I'm sure it's pretty stale. Will check next week.

Also, need to check because this PR might still be directionally right because I think mine is wrong side of the advice (server side adding scope union where your clarification on the expected behaviour was that it should be client side and server should not have to introspect token for active scopes).

Let me get the PR open for review and double check, it has been a while.

**Comment by localden** [comment]
Following up on the SEP side - this PR now has a spec home in [SEP-2350](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2350) (merged), which formalizes client-side scope accumulation. Can we link it from the PR description and the `mergeScopes` JSDoc so the next person knows why this exists?

One gap I noticed cross-checking against the SEP: `_scope` only accumulates from `WWW-Authenticate` headers. The spec says the client should union with "previously requested" scopes, which includes whatever the first auth round derived from PRM `scopes_supported` or `clientMetadata.scope` (the SEP-835 fallback at `auth.ts:669`). Right now that resolved scope never makes it back into `_scope`, so:

1. First auth resolves to `[a, b, c]` from PRM, `_scope` stays `undefined`
2. Server challenges with `scope="d"` per RFC 6750 section 3.1
3. `mergeScopes(undefined, "d")` returns `"d"`, re-auth requests only `d`, drops `a b c`

That's the same loss the PR is fixing for the challenge-to-challenge case, just one hop earlier. IMO the smallest fix is having `auth()` surface `resolvedScope` so the transport can seed `_scope` after the first successful auth. Alternatively we merge `provider.clientMetadata.scope` into the accumulator on construction, but that misses the PRM-derived case.
